lua/websocket/server.lua

```diff
@@ -312,15 +294,163 @@ function SERVER:Think()
 	end
 
 	for i = 1, self.numconnections do
-		self.connections[i]:Think()
-		if (self.connections[i]:GetState() == CLOSED) or (self.connections[i].socket == nil) then
+		local connection = self.connections[i]
+		connection:Think()
+		if connection:GetState() == CLOSED or connection.socket == nil then
 			remove(self.connections, i)
 			self.numconnections = self.numconnections - 1
 			i = i - 1
 		end
 	end
 end
 
+local function clean(self, wasCleanExit, code, reason)
+	if not wasCleanExit and self.errorcallback ~= nil then
+		self.errorcallback(self, "websocket close error: " .. tostring(reason))
+	end
+
+	self:ShutdownInternal(code, reason)
+
+	return false, reason or "closed", wasCleanExit, code
+end
+
+function CONNECTION:ReadFrame()
+	-- https://github.com/lipp/lua-websockets/blob/cafb5874473ab9f54e734c123b18fb059801e4d5/src/websocket/sync.lua#L8-L75
+	--[[
+		Copyright (c) 2012 by Gerhard Lipp <gelipp@gmail.com>
+
+		Permission is hereby granted, free of charge, to any person obtaining a copy
+		of this software and associated documentation files (the "Software"), to deal
+		in the Software without restriction, including without limitation the rights
+		to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+		copies of the Software, and to permit persons to whom the Software is
+		furnished to do so, subject to the following conditions:
+
+		The above copyright notice and this permission notice shall be included in
+		all copies or substantial portions of the Software.
+
+		THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+		IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+		FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+		AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+		LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+		OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+		THE SOFTWARE.
+	]]
+
+	if self.is_closing then
+		return false, "closing"
+	end
+
+	local messages = {}
+
+	local startOpcode
+	local readFrames
+	local bytesToRead = 2
+	local encodedMsg = ""
+
+	while true do
+		local chunk, recErr = self:ReceiveEx(bytesToRead - #encodedMsg)
+		if chunk == nil then
+			return clean(self, false, 1006, recErr)
+		end
+
+		encodedMsg = encodedMsg .. chunk
+
+		local decoded, length, opcode, masked, fin, mask, headerLen = DecodeHeader(encodedMsg)
+		if decoded then
+			chunk, recErr = self:ReceiveEx(length - (#encodedMsg - headerLen))
+
+			if #encodedMsg > headerLen then
+				chunk = encodedMsg:sub(headerLen + 1, -1) .. chunk
+			end
+
+			if masked then
+				chunk = XORMask(chunk, mask)
+			end
+
+			assert(not recErr, "ReadFrame() -> ReceiveEx() error")
+			if opcode == frame.CLOSE then
+				if not self.is_closing then
+					self.is_closing = true
```

Comment:
🙁 `self.is_closing` is set to true inside `ShutdownInternal`. Since it now no longer executes fully if `is_closing` is true, we should probably not set it to true outside that function anymore.